tests/pocket_test.cpp

```diff
@@ -1330,6 +1344,95 @@ TEST_CASE( "character best pocket", "[pocket][character][best]" )
+    item backpack = here.add_item( they.pos(), item( "backpack_hiking" ) );
+    // Rope and backpack are containers, both capable of holding the M4
+    REQUIRE( rope.is_container() );
+    REQUIRE( backpack.is_container() );
+    REQUIRE( rope.can_contain( m4a1 ).success() );
+    REQUIRE( backpack.can_contain( m4a1 ).success() );
+    REQUIRE( backpack.can_contain( rope ).success() );
+
+    // Wearing the backpack, the best pocket for the m4 should be the backpack
+
+    GIVEN( "avatar is wearing only a hiking backpack" ) {
+        they.worn.clear();
+        they.wear_item( backpack );
+
+        // Avatar does not currently possess this rope item
+        REQUIRE_FALSE( they.has_item( rope ) );
+        // This rope item is not in the backpack
+        REQUIRE_FALSE( backpack.has_item( rope ) );
+        // In fact no rope-type item is in the backpack
+        REQUIRE_FALSE( container_has_item_type( backpack, rope ) );
+
+        // All put_in tests are PASSING
+        WHEN( "using put_in to put a rope directly into the backpack" ) {
+            REQUIRE( backpack.put_in( rope, item_pocket::pocket_type::CONTAINER ).success() );
+            they.invalidate_inventory_validity_cache();
+
+            THEN( "the original rope is not in inventory or the backpack" ) {
+                CHECK_FALSE( they.has_item( rope ) );
+                CHECK_FALSE( backpack.has_item( rope ) );
+            }
+            THEN( "a rope-type item is in the backpack" ) {
+                CHECK( container_has_item_type( backpack, rope ) );
+            }
+        }
+
+        // Several i_add tests are FAILING
+        WHEN( "using i_add to put the rope into inventory" ) {
+            // Put the rope in the backpack and get its new reference
+            item &new_rope = they.i_add( rope );
+            they.invalidate_inventory_validity_cache();
+
+            THEN( "the original rope is not in inventory or the backpack" ) {
+                CHECK_FALSE( they.has_item( rope ) );
+                CHECK_FALSE( backpack.has_item( rope ) );
+            }
+            THEN( "a new rope item is in inventory and in the backpack" ) {
+                CHECK( they.has_item( new_rope ) ); // PASSES
+                CHECK( backpack.has_item( new_rope ) ); // FAILS
+            }
+            THEN( "a rope-type item is in the backpack" ) {
+                CHECK( container_has_item_type( backpack, rope ) ); // FAILS
+            }
+        }
+    }
+
+    // FIXME: Most of this section is not working
+    SECTION( "m4 stuff" ) {
+        // Put that rope back in the backpack
+        item new_rope = they.i_add( rope );
+        // Get item_location for m4 on the map
+        item_location m4_loc( map_cursor( they.pos() ), &m4a1 );
+        const drop_locations &thing = { std::make_pair( m4_loc, 1 ) };
+        CHECK_FALSE( backpack.has_item( m4a1 ) );
+
+        // Now pick up the M4
+        they.pick_up( thing );
```

Comment:
So the main issue here is that this only sets up an activity. It also needs to be executed or nothing won't happen. Dunno what's the way to do that in tests, though.

Reply:
I tried `they.activity.do_turn( they )` but that wasn't it...

Reply:
Keep in mind that #54321 isn't merged yet, so the M4 still goes in the rope and the rope with M4 still drops to the ground. (Which also again invalidates the rope reference.)

I also found `process_activity()` in `player_helper.h`, which loops `do_turn` until the activity is over. Using that combined with #54321 passes all checks.

Reply:
Confirmed, thanks. I guess there is no way to get the new rope reference in this case, so without that I won't be able to do any kind of negative test.